text/0000-scoped-custom-elements.md

```diff
@@ -0,0 +1,109 @@
+---
+title: Scoped custom elements
+status: DRAFTED
+created_at: 2021-02-12
+updated_at: 2021-02-12
+pr: (leave this empty until the PR is created)
+---
+
+# Scoped custom elements
+
+## Summary
+
+This feature allows to use scoped custom elements in the core platform as described in the WICG proposal: [Scoped Custom Element Registries](https://github.com/WICG/webcomponents/blob/gh-pages/proposals/Scoped-Custom-Element-Registries.md).
+
+## Motivation
+
+MuleSoft has a pluggable application made with custom elements (you can extend it to any cloud having their non-LWC custom elements). This application is shared cross-platform (Anypoint, the core platform, 3rd-party platforms).
+We see a need for this pluggable application (the components) to be shared with our customers in their on/off-core applications.
+It is currently impossible to use custom elements in an on-core application. The other obstacle is including custom elements in an LWC package, which is scoped by nature. The custom elements should be able to work with different packages. 
+Finally, different customers can use different version of the same components published by us in their applications which under normal circumstances would cause names conflicts.
+
+This is where implementation of scoped custom element registries can solve all of the above problems.
+
+1. In introduces custom elements inside the core platform
+2. LWC based packages create their own registry and passes the reference to it to the component (see below in detailed design section)
+3. Since packages uses custom registries there is no chance for name collision or a component hijacking.
+
+## Detailed design
+
+An LWC package defines own custom elements registry and keeps the reference to be passed to the components.
+
+`registry.js`
+
+```javascript
+export const registry = new CustomElementRegistry();
+```
+
+The custom element used in an LWC has to support a way of passing the registry reference to the shadow root creation. This would require the included custom element to support a static property on the class that keeps the reference to the registry. When the element is being created it uses this reference to pass it to the shadow root creation.
+
+`@scope/custom-element/src/CustomElement.js`
+
+```javascript
+export declare class CustomElement extends HTMLElement {
+  constructor() {
+    const registry = CustomElement.registry;
+    this.attachShadow({mode: 'open', registry});
+  }
+}
+```
+
+The LWC component imports the custom element, passes the handler to the custom element, and registers it.
+
+```javascript
+import { LightningElement } from 'lwc';
+import { CustomElement } from '@scope/custom-element';
+import { registry } from './registry.js';
+CustomElement.registry = registry;
```

Comment:
This is confusing. If the registry is defined manually via a proprietary API (static setter Registry on LightningElement), what should happen with the html file? and the implicit resolution used in LWC today? who wins?